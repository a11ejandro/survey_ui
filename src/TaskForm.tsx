import * as React from 'react'
import TextField from '@material-ui/core/TextField'
import Grid from '@material-ui/core/Grid'
import Button from '@material-ui/core/Button'
import { Task } from './module'

interface State {
  page: number,
  per_page: number
}

interface Props { createdCallback: (task: Task) => void }

export default class TaskForm extends React.Component<Props, State> {
  public constructor(props:Props) {
    super(props)
    this.handleChange = this.handleChange.bind(this)
    this.submit = this.submit.bind(this)

    this.state = {
      page: 0,
      per_page: 0
    }
  }

  public handleChange = (name: string) => (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value } = event.target as HTMLInputElement
    this.setState((prevState) => ({
      ...prevState,
      [name]: value
    }))
  }

  public submit (event: React.SyntheticEvent) {
    event.preventDefault()
    const task = { page: this.state.page, per_page: this.state.per_page} as Task

    this.props.createdCallback(task)
  }

  public render () {
    const {page, per_page} = this.state
    return (
      <Grid container spacing={24}>
        <Grid item xs={4}>
          <TextField
            id='page'
            label="Page"
            value={page}
            onChange={this.handleChange('page')}
            margin="normal"
          />
        </Grid>
        <Grid item xs={4}>
          <TextField
            id='per_page'
            label="Per Page"
            value={per_page}
            onChange={this.handleChange('per_page')}
            margin="normal"
          />
        </Grid>
        <Grid item xs={4}>
          <br/>
          <Button
            id="submit_task"
            variant="outlined"
            color="primary"
            onClick={this.submit}>
            Submit
          </Button>
        </Grid>
      </Grid>
    )
  }
}
