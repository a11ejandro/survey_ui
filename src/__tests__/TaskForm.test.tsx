import * as React from 'react'
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme'

import TaskForm from '../TaskForm'

const createTask= jest.fn()

let mountedComponent: any
const component = () => {
  if (!mountedComponent) {
    mountedComponent = shallow(<TaskForm createdCallback={createTask}/>)
  }
  return mountedComponent
}

it ('renders correctly', () => {
  expect(toJson(component())).toMatchSnapshot()
})

describe ('inputs', () => {
  it ('is initialized with zero values', () => {
    expect(component().state().page).toBe(0)
    expect(component().state().per_page).toBe(0)
  })


  it('connects page input to state', () => {
    const input = component().find('TextField#page').first()
    const event = { target: { value: 9000 }}
    input.simulate('change', event)
    expect(component().state().page).toBe(9000)
  })

  it('connects per page input to state', () => {
    const input = component().find('TextField#per_page').first()
    const event = { target: { value: 2061 }}
    input.simulate('change', event)
    expect(component().state().per_page).toBe(2061)
  })
})

describe ('task creation', () => {
  it ('calls callback with current page and per_page', () => {
    component().setState({
      page: 22,
      per_page: 22
    })

    const button = component().find('#submit_task')
    button.simulate('click', { preventDefault: () => false })
    expect(createTask).toBeCalledWith({page: 22, per_page: 22})
  })
})
