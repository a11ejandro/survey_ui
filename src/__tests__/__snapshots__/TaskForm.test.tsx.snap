// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<WithStyles(Grid)
  container={true}
  spacing={24}
>
  <WithStyles(Grid)
    item={true}
    xs={4}
  >
    <TextField
      id="page"
      label="Page"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={false}
      value={0}
      variant="standard"
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={4}
  >
    <TextField
      id="per_page"
      label="Per Page"
      margin="normal"
      onChange={[Function]}
      required={false}
      select={false}
      value={0}
      variant="standard"
    />
  </WithStyles(Grid)>
  <WithStyles(Grid)
    item={true}
    xs={4}
  >
    <br />
    <WithStyles(Button)
      color="primary"
      id="submit_task"
      onClick={[Function]}
      variant="outlined"
    >
      Submit
    </WithStyles(Button)>
  </WithStyles(Grid)>
</WithStyles(Grid)>
`;
